bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/BKTClientImpl.kt

```diff
@@ -0,0 +1,169 @@
+import java.util.concurrent.Future
+import java.util.concurrent.ScheduledExecutorService
+
+internal class BKTClientImpl(
+  private val context: Context,
+  private val config: BKTConfig,
+  user: BKTUser,
+  private val userHolder: UserHolder = UserHolder(user.toUser()),
+  private val executor: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(),
+  internal val component: Component = Component(
+    dataModule = DataModule(
+      application = context.applicationContext as Application,
+      config = config,
+    ),
+    interactorModule = InteractorModule(),
+  ),
+) : BKTClient {
+
+  override fun stringVariation(featureId: String, defaultValue: String): String {
+    return getVariationValue(featureId, defaultValue)
+  }
+
+  override fun intVariation(featureId: String, defaultValue: Int): Int {
+    return getVariationValue(featureId, defaultValue)
+  }
+
+  override fun doubleVariation(featureId: String, defaultValue: Double): Double {
+    return getVariationValue(featureId, defaultValue)
+  }
+
+  override fun booleanVariation(featureId: String, defaultValue: Boolean): Boolean {
+    return getVariationValue(featureId, defaultValue)
+  }
+
+  override fun jsonVariation(featureId: String, defaultValue: JSONObject): JSONObject {
+    return getVariationValue(featureId, defaultValue)
+  }
+
+  override fun track(goalId: String, value: Double) {
+    val user = userHolder.get()
+    val featureTag = config.featureTag
+    executor.execute {
+      component.eventInteractor.trackGoalEvent(
+        featureTag = featureTag,
+        user = user,
+        goalId = goalId,
+        value = value,
+      )
+    }
+  }
+
+  override fun currentUser(): BKTUser {
+    return userHolder.get().toBKTUser()
+  }
+
+  override fun setUserAttributes(attributes: Map<String, String>) {
+    userHolder.update { it.copy(data = attributes) }
+  }
+
+  private fun fetchEvaluationsSync(timeoutMillis: Long?): BKTException? {
+    val result = component.evaluationInteractor.fetch(user = userHolder.get(), timeoutMillis)
+
+    executor.execute {
+      val interactor = component.eventInteractor
+      when (result) {
+        is GetEvaluationsResult.Success -> {
+          interactor.trackFetchEvaluationsSuccess(
+            featureTag = result.featureTag,
+            mills = result.millis,
+            sizeByte = result.sizeByte,
+          )
+        }
+        is GetEvaluationsResult.Failure -> {
+          interactor.trackFetchEvaluationsFailure(
+            featureTag = result.featureTag,
+            error = result.error,
+          )
+        }
+      }
+    }
+
+    return when (result) {
+      is GetEvaluationsResult.Success -> null
+      is GetEvaluationsResult.Failure -> result.error
+    }
+  }
+
+  override fun fetchEvaluations(timeoutMillis: Long?): Future<BKTException?> {
+    return executor.submit<BKTException?> {
+      fetchEvaluationsSync(timeoutMillis)
+    }
+  }
+
+  // should we return Future?
```

Comment:
I think it would be better to return `Future` if someone needs to ensure that events are sent before doing something else.

Reply:
thanks, done in [`a740c84` (#68)](https://github.com/ca-dp/bucketeer-android-sdk/pull/68/commits/a740c841d5f14d58665fd7de2604de4f6d1c3075)